HPA metrics specificity improvements

damemi opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Improves available specificity for HPA metrics by adding metric selector fields for metrics of Pods and Objects. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Implements this KEP: https://github.com/kubernetes/community/pull/2055

**Special notes for your reviewer**:
Need to add/update tests?

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Introduces autoscaling/v2beta2 and custom_metrics/v1beta2, which implement metric selectors for Object and Pods metrics, as well as allowing AverageValue targets on Objects, similar to External metrics.
```

/assign @DirectXMan12

DirectXMan12 commented:
For the APIServices int test, you need to enable the API version in PKG/master/master.go

DirectXMan12 commented:
Make sure we have validation on the metric target

DirectXMan12 commented:
you missed some cases of switching to `scheme.ParameterCodec` in the custom metrics client.

DirectXMan12 commented:
conversions aren't quite right -- you don't properly populate the `type` field in the metric target, it looks like

DirectXMan12 commented on pkg/apis/autoscaling/v2beta1/conversion.go:
```diff
@@ -0,0 +1,399 @@
+	return nil
+}
+
+func Convert_autoscaling_MetricTarget_To_v2beta1_CrossVersionObjectReference(in *autoscaling.MetricTarget, out *autoscalingv2beta1.CrossVersionObjectReference, s conversion.Scope) error {
```
why are these nil?  If they're not needed, don't implement them

damemi replied:
That's what I thought, but without them defined as nil they get generated for some reason:

```
# k8s.io/kubernetes/pkg/apis/autoscaling/v2beta1
pkg/apis/autoscaling/v2beta1/zz_generated.conversion.go:507:12: undefined: Convert_v2beta1_CrossVersionObjectReference_To_autoscaling_MetricTarget
pkg/apis/autoscaling/v2beta1/zz_generated.conversion.go:519:12: undefined: Convert_autoscaling_MetricTarget_To_v2beta1_CrossVersionObjectReference
!!! [0530 21:24:29] Call tree:
!!! [0530 21:24:29]  1: /home/mdame/go/src/k8s.io/kubernetes/hack/lib/golang.sh:595 kube::golang::build_binaries_for_platform(...)
!!! [0530 21:24:29]  2: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0530 21:24:29] Call tree:
!!! [0530 21:24:29]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0530 21:24:29] Call tree:
!!! [0530 21:24:29]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make: *** [Makefile:92: all] Error 1
```

I think it has to do with the fact that `Target` in `v2beta1` is a `CrossVersionObjectReference`, but in `v2beta2` we renamed that to `DescribedObject` and created a *new* field called `Target` that is a `MetricTarget`. So it's seeing Target=Target and trying to auto-convert CrossVersionObjectReference->MetricTarget and telling me I need to manually convert it (which I can't)

If I'm wrong, or there's a way around that, I can make the change (or I can add a comment explaining these functions, which I probably should have done first)

DirectXMan12 commented on pkg/apis/autoscaling/v2beta1/conversion.go:
```diff
@@ -0,0 +1,399 @@
+func Convert_autoscaling_ResourceMetricSource_To_v2beta1_ResourceMetricSource(in *autoscaling.ResourceMetricSource, out *autoscalingv2beta1.ResourceMetricSource, s conversion.Scope) error {
+	out.Name = v1.ResourceName(in.Name)
+	out.TargetAverageUtilization = in.Target.AverageUtilization
+	out.TargetAverageValue = in.Target.AverageValue
```
you need to persist selector, too.

damemi replied:
`ResourceMetricSource` doesn't have a selector in the new version or v2beta1, was that overlooked?

DirectXMan12 replied:
Yeah, you'll need to add a selector field.  New API versions are weird.

DirectXMan12 replied:
(to the old version)

damemi replied:
My point is, there isn't a selector field in the new version to persist to the old version. The KEP only adds `Target` to Resource metrics. I'm assuming that was intentional. Did you want to add metric selectors to Resource metrics as well?

damemi commented:
@DirectXMan12 made the changes in your comments, updated validations and added some tests for new fields, updated conversions, and tests pass locally. I had some questions on your other comments that I'm not sure about. Also need to add tests for the new apis but I'm not exactly sure where, I can do that today

DirectXMan12 commented on pkg/master/master.go:
```diff
@@ -35,6 +35,7 @@ import (
 	authorizationapiv1beta1 "k8s.io/api/authorization/v1beta1"
 	autoscalingapiv1 "k8s.io/api/autoscaling/v1"
 	autoscalingapiv2beta1 "k8s.io/api/autoscaling/v2beta1"
+	autoscalingapiv2beta2 "k8s.io/api/autoscaling/v2beta1"
```
you miss-copied ;-)

It says `/autoscaling/v2beta1, you need `/autoscaling/v2beta2`

damemi replied:
Ah! Technically I copied it perfectly, it's the changing it that I messed up. Thanks for catching that

damemi commented:
/retest

DirectXMan12 commented:
@damemi please squash commits down to a reasonable set (doesn't have to be one, but each commit should represent some distinct chunk of work).

damemi commented:
Made changes and squashed to a single commit

damemi commented:
/retest

damemi commented:
/retest

DirectXMan12 commented:
@lavalamp can you or something else on @kubernetes/api-reviewers take a quick peek to make sure this is copacetic, and then approve?

damemi commented:
/retest

damemi commented:
/retest

DirectXMan12 commented:
/assign @lavalamp for visibility

damemi commented:
/retest

damemi commented:
/retest

damemi commented:
/retest

damemi commented:
/retest

damemi commented:
/retest

damemi commented:
/retest

damemi commented:
/retest

smarterclayton commented:
Please separate out api changes into their own commits.  This is a lot harder to review.

smarterclayton commented on staging/src/k8s.io/api/autoscaling/v1/types.go:
```diff
@@ -223,6 +226,9 @@ type PodsMetricSource struct {
 	TargetAverageValue resource.Quantity `json:"targetAverageValue" protobuf:"bytes,2,name=targetAverageValue"`
+
+	Selector string             `json:"selector,omitempty" protobuf:"bytes,3,name=selector"`
+	Value    *resource.Quantity `json:"value,omitempty" protobuf:"bytes,4,name=value"`
```
You need descriptions of how these are used.

smarterclayton commented on staging/src/k8s.io/api/autoscaling/v1/types.go:
```diff
@@ -267,6 +273,16 @@ type ExternalMetricSource struct {
+type MetricTargetType string
+
+var (
+	UtilizationMetricType  MetricTargetType = "Utilization"
```
These should be `const` and they should have Godoc.

damemi committed:
Autoscaling v2beta2 and custom metrics v1beta2 APIs

damemi committed:
Update autoscaling conversion and validation for v2beta2 inclusion

damemi committed:
Implement autoscaling/v2beta2 features in HPA controller

damemi committed:
Update metrics API to include autoscaling/v2beta2 changes

damemi committed:
Update printers to include changes from autoscaling/v2beta2

damemi committed:
Add autoscaling/v2beta2 and custom_metrics/v1beta2 to necessary files

damemi committed:
Generate files and modifications for autoscaling/v2beta2 and custom_metrics/v1beta2

damemi commented:
/retest

damemi commented:
@smarterclayton thanks, broke it out into what I hope is a sensible order of commits

smarterclayton commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -0,0 +1,361 @@
+// any Kubernetes object (for example length of queue in cloud
+// messaging service, or QPS from loadbalancer running outside of cluster).
+type ExternalMetricSource struct {
+	Metric MetricIdent  `json:"metric" protobuf:"bytes,1,name=metric"`
```
Both of these need comments.

smarterclayton commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -0,0 +1,361 @@
+}
+
+// MetricIdent defines the name and optionally selector for a metric
+type MetricIdent struct {
```
This name is too short.  Lengthen it to two complete words - `MetricIdentifier` or `MetricReference`.

smarterclayton commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -0,0 +1,361 @@
+	// selector is the selector for the given metric
+	// it is the string-encoded form of a standard kubernetes label selector
+	// +optional
+	Selector string `json:"selector,omitempty" protobuf:"bytes,2,name=selector"`
```
This was previously a LabelSelector.  Why did it change?  We use LabelSelector consistently elsewhere, there needs to be a reason to deviate.

DirectXMan12 replied:
hmm... I think this was a string in one form of Scale, and got copied over (and maybe internally at some point).  Go ahead and switch it over to a LabelSelector.

smarterclayton commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -0,0 +1,361 @@
+type MetricTarget struct {
+	// Type represents whether the metric type is Utilization, Value, or AverageValue
+	Type         MetricTargetType   `json:"type" protobuf:"bytes,1,name=type"`
+	Value        *resource.Quantity `json:"value,omitempty" protobuf:"bytes,2,name=value"`
```
Godoc for value.

smarterclayton commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -0,0 +1,361 @@
+	// Type represents whether the metric type is Utilization, Value, or AverageValue
+	Type         MetricTargetType   `json:"type" protobuf:"bytes,1,name=type"`
+	Value        *resource.Quantity `json:"value,omitempty" protobuf:"bytes,2,name=value"`
+	AverageValue *resource.Quantity `json:"averageValue,omitempty" protobuf:"bytes,3,name=averageValue"`
```
Godoc for average value.

smarterclayton commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -0,0 +1,361 @@
+
+// MetricTarget defines the target value, average value, or average utilization of a specific metric
+type MetricTarget struct {
+	// Type represents whether the metric type is Utilization, Value, or AverageValue
```
Godoc comments on external APIs should be camelCase, like the keys a user would see in JSON.

smarterclayton commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -0,0 +1,361 @@
+	Value        *resource.Quantity `json:"value,omitempty" protobuf:"bytes,2,name=value"`
+	AverageValue *resource.Quantity `json:"averageValue,omitempty" protobuf:"bytes,3,name=averageValue"`
+
+	// Currently only valid for Resource metric source type
```
This needs more of a description.

smarterclayton commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -0,0 +1,361 @@
+
+// Percentage is an integer internally (like before), but
+// serializes as "<number>%" to avoid user confusion.
+type Percentage = int32
```
This is clever and novel, but I'm not sold.  Where is the serialization actually defined?  What happens if the user omits the percentage?  What if they provide `0`?   Is `%` by itself allowed?  How are you reporting errors back to the user.

I'm going to need some more justification before we introduce something like this.  And it will need a lot more documentation.

DirectXMan12 replied:
I don't think the code every got implemented for this -- it was supposed to be yanked.  I requested API review on it because I wasn't sure if it was legit.

My hope with the original proposal was that it would make it clearer to users what this field was, and what the units were, because as it stands, it's a bit confusing.  It's no big deal with it gets removed, but I would like to consider it eventually.

smarterclayton commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -0,0 +1,361 @@
+type MetricTargetType string
+
+var (
+	UtilizationMetricType  MetricTargetType = "Utilization"
```
Needs godoc.

smarterclayton commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -0,0 +1,361 @@
+// ObjectMetricStatus indicates the current value of a metric describing a
+// kubernetes object (for example, hits-per-second on an Ingress object).
+type ObjectMetricStatus struct {
+	Metric  MetricIdent       `json:"metric" protobuf:"bytes,1,name=metric"`
```
Godoc

smarterclayton commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -0,0 +1,361 @@
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// HorizontalPodAutoscaler is the configuration for a horizontal pod
```
Please group all structs of a related type together - by convention we do:

1. core object (HPA)
2. spec (HPASpec)
3. sub types of spec (...)
4. status (HPAStatus)
5. sub types of status if not already defined.

damemi commented:
Still need to update the LabelSelector, but pushing now. Would like to let tests run and make sure I didn't miss anything in all the renaming

damemi commented:
/retest

damemi commented:
@smarterclayton applied your feedback so far, is there anything else that needs to be updated?

zparnold commented:
/milestone v1.12
/meow

DirectXMan12 commented:
o_O what sorcery is this?  A bot that provides cat pictures (but seriously, maybe just provide links, or make sure they're smaller).

zparnold commented:
Yea, sorry. :) I wanted to try out some of the other Prow commands, and it
went wrong. I wanted to delete it, but alas...there is no undo here. :(

On Fri, Jul 13, 2018 at 1:47 PM Solly Ross <notifications@github.com> wrote:

> o_O what sorcery is this? A bot that provides cat pictures (but seriously,
> maybe just provide links, or make sure they're smaller).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/64097#issuecomment-404949316>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE81SNExAcuTCh-7RRonX1zELCZOIXaLks5uGQdTgaJpZM4UHEV3>
> .
>

krzyzacy commented:
:-) @DirectXMan12 you don't like it?

BenTheElder commented:
perhaps we should discuss this in a test-infra issue, but `/meow` is roughly equivalent to a user commenting with a cat image themselves.. avoiding disruptiveness is kinda up to the user. we mostly use it when approving. `/lgtm\n/meow space` makes for a great review 🙃

zparnold commented:
/milestone v1.12

zparnold commented:
/sig autoscaling
/priority important-soon
/kind feature

DirectXMan12 commented:
> perhaps we should discuss this in a test-infra issue, but /meow is roughly equivalent to a user commenting with a cat image themselves.. avoiding disruptiveness is kinda up to the user. we mostly use it when approving. /lgtm\n/meow space makes for a great review upside_down_face

I'll be honest, I didn't actually see the `/meow`, thought it was in response to the milestone commit.

I do like cat pictures, :smile_cat: (a *lot*), but I prefer when they don't take up 2-3 comments worth of space.  Maybe I'm being curmudgeonly.  Perhaps put it in a `<details></details>` block, or grab thumbnails instead of full-size images?

/me shrugs

DirectXMan12 commented:
/status approved-for-milestone

damemi commented:
@smarterclayton did you have any more feedback you'd like to add to this?

smarterclayton commented on staging/src/k8s.io/api/autoscaling/v1/types.go:
```diff
@@ -223,6 +232,14 @@ type PodsMetricSource struct {
 	// metric across all relevant pods (as a quantity)
 	TargetAverageValue resource.Quantity `json:"targetAverageValue" protobuf:"bytes,2,name=targetAverageValue"`
+
+	// selector is the selector for the given metric
```
These need to be sentences - newlines are not considered significant and will cause this to be run on output.

damemi replied:
Done

smarterclayton commented on staging/src/k8s.io/api/autoscaling/v2beta1/types.go:
```diff
@@ -294,6 +320,14 @@ type PodsMetricStatus struct {
+	Selector *metav1.LabelSelector `json:"selector,omitempty" protobuf:"bytes,3,name=selector"`
+	// value is the current value of the metric (as a quantity)
+	// +optional
+	Value *resource.Quantity `json:"value,omitempty" protobuf:"bytes,4,name=value"`
```
You need to explain how this is different from CurrentAverageValue.  Reading the godoc, I have no idea.

damemi replied:
This was actually added in error -- `averageValue` is sufficient for pods metrics. Thanks for catching, removed

smarterclayton commented on pkg/apis/autoscaling/v2beta2/defaults.go:
```diff
@@ -0,0 +1,51 @@
+		obj.Spec.MinReplicas = &minReplicas
+	}
+
+	if len(obj.Spec.Metrics) == 0 {
```
Is it ever valid to have no metrics?  I think you need to add as a comment on the Metrics field in v2beta2 about the default.  I.e "if not set, the default metric will be X% CPU utilization" (fill in the X)

damemi replied:
This is just following the defaults set by previous versions of the autoscaling API. Added a comment to the API about the default percentage (currently 80)

smarterclayton commented on staging/src/k8s.io/metrics/pkg/apis/custom_metrics/v1beta1/types.go:
```diff
@@ -56,6 +56,12 @@ type MetricValue struct {
+
+	// selector represents the label selector that could be used to select
+	// this metric, and will generally just be the selector passed in to
+	// the query used to fetch this metric.
```
Describe what happens if this is empty.

damemi replied:
Done

smarterclayton commented on staging/src/k8s.io/api/autoscaling/v1/types.go:
```diff
@@ -223,6 +231,10 @@ type PodsMetricSource struct {
 	// metric across all relevant pods (as a quantity)
 	TargetAverageValue resource.Quantity `json:"targetAverageValue" protobuf:"bytes,2,name=targetAverageValue"`
+
+	// selector is the string-encoded form of a standard kubernetes label selector for the given metric
```
Explain what selector means to the user in this context.  Why would they set it, what happens if it is empty, etc.

damemi replied:
Done (also in other areas this appears-- v2beta1 and v2beta2)

smarterclayton commented on staging/src/k8s.io/api/autoscaling/v1/types.go:
```diff
@@ -344,6 +356,14 @@ type ObjectMetricStatus struct {
 	// currentValue is the current value of the metric (as a quantity).
 	CurrentValue resource.Quantity `json:"currentValue" protobuf:"bytes,3,name=currentValue"`
+
+	// selector is the string-encoded form of a standard kubernetes label selector for the given metric
```
Explain what selector means to the user in this context.  Why would they set it, what happens if it is empty, etc.

smarterclayton commented on staging/src/k8s.io/api/autoscaling/v1/types.go:
```diff
@@ -211,6 +211,14 @@ type ObjectMetricSource struct {
 	// targetValue is the target value of the metric (as a quantity).
 	TargetValue resource.Quantity `json:"targetValue" protobuf:"bytes,3,name=targetValue"`
+
+	// selector is the string-encoded form of a standard kubernetes label selector for the given metric
```
Explain what selector means to the user in this context.  Why would they set it, what happens if it is empty, etc.

damemi replied:
@smarterclayton updated with your feedback, is there anything else that stands out?

damemi commented:
/retest

krzyzacy commented:
/retest

damemi commented:
/test pull-kubernetes-integration

damemi commented:
/retest

damemi committed:
Add dry run test for hpa v2beta2

smarterclayton commented:
/approve

DirectXMan12 commented:
/lgtm
